packages/create-subscription/src/createSubscription.js

```diff
@@ -0,0 +1,153 @@
+    // They should call the callback() instead when a subscription changes.
+    subscribe: (
+      source: Property,
+      callback: (value: Value | void) => void,
+    ) => Unsubscribe | CannotUnsubscribe,
+  |}>,
+): React$ComponentType<{
+  children: (value: Value | void) => React$Node,
+  source: any,
+}> {
+  const {getValue, subscribe} = config;
+
+  warning(
+    typeof getValue === 'function',
+    'Subscription must specify a getValue function',
+  );
+  warning(
+    typeof subscribe === 'function',
+    'Subscription must specify a subscribe function',
+  );
+
+  type Props = {
+    children: (value: Value) => React$Element<any>,
+    source: any,
+  };
+  type State = {
+    source: Property,
+    unsubscribeContainer: {
+      unsubscribe?: Unsubscribe | CannotUnsubscribe,
+    },
+    value: Value | void,
+  };
+
+  // Reference: https://gist.github.com/bvaughn/d569177d70b50b58bff69c3c4a5353f3
+  class Subscription extends React.Component<Props, State> {
+    state: State = {
+      source: this.props.source,
+      unsubscribeContainer: {},
+      value:
+        this.props.source != null ? getValue(this.props.source) : undefined,
+    };
+
+    static getDerivedStateFromProps(nextProps, prevState) {
+      if (nextProps.source !== prevState.source) {
+        return {
+          source: nextProps.source,
+          unsubscribeContainer: {},
+          value:
+            nextProps.source != null ? getValue(nextProps.source) : undefined,
+        };
+      }
+
+      return null;
+    }
+
+    componentDidMount() {
+      this.subscribe();
+    }
+
+    componentDidUpdate(prevProps, prevState) {
+      if (this.state.source !== prevState.source) {
+        this.unsubscribe(prevState);
+        this.subscribe();
+      }
+    }
+
+    componentWillUnmount() {
+      this.unsubscribe(this.state);
+    }
+
+    render() {
+      return this.props.children(this.state.value);
+    }
+
+    subscribe() {
+      const {source} = this.state;
+      if (source != null) {
+        const callback = (value: Value | void) => {
+          this.setState(state => {
+            // If the value is the same, skip the unnecessary state update.
+            if (value === state.value) {
+              return null;
+            }
+
+            // If this event belongs to an old or uncommitted data source, ignore it.
+            if (source !== state.source) {
+              return null;
+            }
+
+            return {value};
+          });
+        };
+
+        // Store subscription for later (in case it's needed to unsubscribe).
+        // This is safe to do via mutation since:
+        // 1) It does not impact render.
+        // 2) This method will only be called during the "commit" phase.
+        const unsubscribeOrBoolean = subscribe(source, callback);
+
+        invariant(
+          unsubscribeOrBoolean === false ||
+            typeof unsubscribeOrBoolean === 'function',
```

Comment:
I suppose `subscribe` isn't as hot a path as, say, the render phase, but could we change this to a warning?

Reply:
Maybe you should always have to return an unsubscribe function? If it's not supported, you can return a noop. By supporting a boolean, it implies that either option is equally fine, whereas you should always unsubscribe if possible.

Reply:
I think warnings might be too easy to overlook. I could add one, but I think we'd also want an error later (when unsubscribing) which happens just as often, so what's the point?

Reply:
Because warnings can be stripped out at runtime. If the code is correct, there's no overhead. invariants add runtime overhead regardless of whether the code is correct. It's the same reason prop-types uses warnings instead of invariants.

Reply:
Wouldn't we want to keep a meaningful error message in either case?

I could change this to a warning if you feel strongly that's the better way to go.